pages/user-interfaces.html

```diff
@@ -55,82 +43,98 @@
+use <a href="https://jscoq.github.io/">JsCoq</a> which loads Coq
+entirely in <strong>your browser</strong>.  However, it is limited to conduct
+actual projects: e.g., it lacks access to the VM and native computing
+machineries of Coq, and running in the browser is slower.
 </p>
 
 </div>
 
 <div class="frameworklinks">
 <ul>
-<li><a class="extlink" href="https://github.com/coq-community/vscoq">VsCoq (VS Code)</a></li>
-<li><a class="extlink" href="https://github.com/ejgallego/coq-lsp">coq-lsp (VS Code)</a></li>
+<li><a class="extlink" href="https://github.com/coq-community/vscoq/tree/vscoq1">VsCoq 1 (VS Code)</a></li>
 <li><a class="extlink" href="https://proofgeneral.github.io/">Proof General (Emacs)</a></li>
 <li><a class="extlink" href="https://github.com/whonore/Coqtail">Coqtail (Vim)</a></li>
+<li><a href="refman/practical-tools/coqide.html">CoqIDE</a></li>
+<li><a class="extlink" href="https://jscoq.github.io/">JsCoq</a></li>
 </ul>
 </div>
 </div>
 
 <div class="framework">
-<div class="frameworklabel">Standalone interfaces</div>
+<div class="frameworklabel">Towards a modular design of Coq user interfaces based on the LSP protocol</div>
 <div class="frameworkcontent">
-
-<p>
-Alternatively, you can
-use <a href="https://coq.inria.fr/refman/practical-tools/coqide.html">CoqIDE</a>,
-which is developed and distributed alongside Coq. CoqIDE relies
-on <a href="https://github.com/coq/coq/blob/master/dev/doc/xml-protocol.md">Coq's
-XML protocol for IDEs</a> and implements all of the features provided
-by this protocol, meaning in particular asynchronous evaluation.
-CoqIDE comes with standard basic Emacs-like bindings
-and with some specific Coq-related bindings but, as an editor, it is
-not as complete as a general-purpose editor could be (for instance it
-does not have automatic indentation).
-</p>
-
-<p>
-As a way to try Coq without installing anything, you can
-use <a href="https://jscoq.github.io/">JsCoq</a>.  JsCoq loads Coq
-entirely in your browser.  However, it is too limited to conduct
-actual projects: it lacks access to the VM and native computing
-machineries of Coq, and may hit out-of-memory and stack-overflow
-failures quicker than native versions of Coq.
-</p>
-
+  The <a href="https://microsoft.github.io/language-server-protocol">Language
+        Server Protocol</a> (LSP) is a common protocol for user
+        interfaces. Two interfaces, each composed of a VS Code
+        client and of a Coq server communicating over a proof-assistant-specific
+        extension of LSP are at an experimental stage of development:
+  <ul>
+    <li><a href="https://github.com/coq-community/vscoq">VsCoq 2</a></li>
+    <li><a href="https://github.com/ejgallego/coq-lsp">coq-lsp</a></li>
+  </ul>
+  <p>
+  Both provides the ability of continuous and asynchronous document
+  checking as well as position-based document navigation (an
+  interaction model introduced in Wenzel's Isabelle/JEdit which
```

Comment:
It might be worth mentioning that VsCoq2 also supports switching to a legacy-style mode of interaction if it doesn't make the text too long.

Reply:
What about "Both provides the extra ability ..."? (In the case of coq-lsp, afaiu, the server supports evaluation to a chosen position but the client does not - yet? - give access to it.)

Reply:
IMO, what the users most care about is what is accessible for the client. But OK to me if you do not want to make a comparison here.

Reply:
IIUC, Emilio is not fully convincent that it is worth to support the PG mode and this is why the client is bridled. I temporarily wrote "Both servers provides the extra ability" until it is clear what is the intent of coq-lsp developers.

Reply:
OK for now, but if coq-lsp doesn't support the legacy mode on purpose, it will be important to make the distinction when we start to advertise these two options more.

Reply:
> I temporarily wrote "Both servers provides the extra ability" until it is clear what is the intent of coq-lsp developers.

You didn't push that change AFAICS.

Reply:
It is done.

Reply:
I still don't see it. But I'm fine with the current state FWIW.

Reply:
Ah OK, my mistake. I understand what you meant now.

Reply:
> OK for now, but if coq-lsp doesn't support the legacy mode on purpose, it will be important to make the distinction when we start to advertise these two options more.

Technically coq-lsp does support checking to a point, and indeed it uses internally in an essential way (for example the way click to show goals works today is exactly like that)

The reason I didn't put an option for that is that 0 users do prefer the old mode, which has pretty bad latency properties. However there is a chicken and egg design problem here, in the sense that once you are used to the more eager mode, you never want to go back. Design in this sense should not work IMO by having the user do these choices. (Neither lean of isabelle allow that for example). If something, coq-lsp can read the power profile and be more eager / conservative, but always you want to do a bit of prefetching as Isabelle does, even in low power mode (which can the properly limited by the DM to have the right CPU / mem bounds)

Anyways we need a better setup for the start of the page than to expose all these very technical things IMHO.

Reply:
>The reason I didn't put an option for that is that 0 users do prefer the old mode

You are always criticizing others when they make claims without evidence or false claims, but then you come and make such strong claims yourself. Do I need to remind you that Clément said explicitly that he didn't like the new mode from his experience with other proof assistants that use it?

Reply:
@Zimmi48 , I'm sorry to hear you think that me pointing out that a claim seems incorrect to me sounds like a critic; it is for sure not my intention to critizise but to point out that once I have reviewed what the code is doing I found different. I'd be happy to hear how could I do better in this sense.

In this case I have discussed with maybe a dozen users about their preferences, (and actually I'd love to run a true study on this), and none of the users did prefer the old mode, when we enabled it the bad latency alone makes a big impact on the usability (sample size here is N=3 tho, so again pretty low).

Note that I took Clément feedback into account, indeed you want to check less for example when you are in low power mode etc..., which IIUC that was a core problem he had with other systems. For example I understood that he'd be fine with the current hint Isabelle uses (and that is planned, but requires a bit more work on the core engine)

I was just trying to clarify, based on what Hugo had asked me in person, what was the status of such a feature in coq-lsp and why we didn't put an option to that. If there is interest in having such a mode I'm happy to add and option to enable it as it is only a one-line change in the engine (as I pointed out to Hugo in person)